### PR TITLE
Add command line interface

### PR DESCRIPTION
**Pull request by kmadsen**
Resolves https://github.com/mapbox/navigation-sdks/issues/542

## Usage

1. build
```
➜  mapbox-java git:(km-services-cli) ./gradlew shadowJar
or
➜  mapbox-java git:(km-services-cli) make build-cli
```
2. commands
```
➜  mapbox-java git:(km-services-cli) java -jar services-cli/build/libs/services-cli-all.jar
Missing required option: f
usage: java -jar services-cli/build/libs/mapboxjava-cli.jar
 -f,--file <arg>   Path to a json file or directory
 -h,--help         Show command line interface usage
```
3. validate json
```
➜  mapbox-java git:(km-services-cli) java -jar services-cli/build/libs/services-cli-all.jar -f services-directions/src/test/resources/directions_v5_waypoint_names.json
[{"filename":"directions_v5_waypoint_names.json","success":true,"converts_back":false}]
```

## Alternatives

The original ticket asks to test json on any previous build of mapbox-java, which is a little complicated with gradle - so here was an alternative to make the CLI in NodeJs https://github.com/mapbox/mapbox-java/pull/1189 - This approach was closed because we would rather build it in kotlin.

## Next steps

**Distributing**
We will want to add this jar to our [mapbox-java releases](https://github.com/mapbox/mapbox-java/releases) as an asset. Notice how [classyshark](https://github.com/google/android-classyshark/releases) does this too. This task is to make it so directions api can [use them in their CI process](https://github.com/mapbox/navigation-sdks/issues/542#issuecomment-696348267).

**Features**
Currently you can make directions api calls [with `make` commands using `curl`](https://github.com/mapbox/mapbox-java/blob/af2493abd19417e7f18249cf1f5c7a0ca4920b1f/Makefile#L54-L209). This project will allow us to move those commands to use the actual mapbox-java implementation (it's not implemented yet, but this project is set up to hopefully make it easy to do so)

Feel free to add your ideas! This is also written in kotlin, so it will be easier to start building our mapbox-java modules in kotling.

### CODE REVIEW

**Comment by kmadsen** [comment]
cc: @zugaldia @Guardiola31337 @langsmith

**Comment by zugaldia** [comment]
This is great @kmadsen - I just gave it a try, no issues. A few random suggestions, in no particular order, some of them could be tracked as tailwork:
1. In addition to validating a JSON, it'd be good if we could set an arbitrary endpoint and execute an actual request like you mention in OP. That way, we could validate, for example, staging endpoints.
1. There's a [handy Gradle plugin](https://github.com/johnrengelman/shadow) called shadow that creates "fat jars". It combines a project's dependency classes and resources into a single output jar. This makes the jar heavier, but self-contained.
1. Building on the item above, I'd document the process to attach the resulting jar with every release.

**Comment by kmadsen** [comment]
> 2. use shadowJar

Yea thanks for the reminder! removes the TODO because they already solved the problems I was dealing with. it's all updated

> 1. That way, we could validate, for example, staging endpoints

Yep agree 👍 ; was planning on making that another change so it can receive separate code review

> 3. I'd document the process to attach the resulting jar with every release.

Agree again 👍 ; Was also wanting to make this tail work, would be nice to make it more automated than documented

**Review by Guardiola31337 — APPROVED** [review]
Great work @kmadsen 

Left a minor comment (not blocking the PR).

Another follow up would be creating a Gradle plugin that runs the `MapboxJavaCli` program.

Also don't forget to cut tickets with the tailwork listed in https://github.com/mapbox/mapbox-java/pull/1198#issuecomment-702519327

File: services-cli/build.gradle [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+        maven { url 'https://plugins.gradle.org/m2' }
+        // To use snapshots
+        //maven { url 'https://oss.jfrog.org/artifactory/oss-snapshot-local/' }
+        jcenter()
+    }
+
+    dependencies {
+        classpath pluginDependencies.buildConfig
+        classpath 'org.jetbrains.kotlin:kotlin-gradle-plugin:1.3.72'
+        classpath 'com.github.jengelman.gradle.plugins:shadow:4.0.4'
```
[the remark this answers is not in the document]
Comment (kmadsen) [reply services-cli/build.gradle]: yeah moved some of them. tried to upgrade gradle but that wasn't totally working yet

File: services-cli/build.gradle [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+
+    // The Apache Software License, Version 2.0
+    // http://www.apache.org/licenses/LICENSE-2.0.txt
+    implementation("commons-cli:commons-cli:1.4")
+
+    // Kotlin
+    implementation("org.jetbrains.kotlin:kotlin-stdlib-jdk8")
+
+    // Testing
+    testImplementation("org.jetbrains.kotlin:kotlin-test-junit")
```
Comment (Guardiola31337) [inline services-cli/build.gradle:47]: We should move these to `gradle/dependencies.gradle` https://github.com/mapbox/mapbox-java/pull/1198#discussion_r500435695

**Pull request merged**